### ISSUE TITLE
Proposal: new architecture/structure for Apache APISIX Documentation

### ISSUE THREAD

**Issue by avinal** [apache/apisix#6461]
# What needs Improvements?

### Missing Introduction page - must have a proper introduction (partial)
  -  What is an API Gateway?
  - What is Apache APISIX?
  - Features in brief
  - Architectural Design (link to architectural design)
  - Katacoda links and other tutorials
  - Videos to get started with
  - Link to contribute to the project (link to CONTRIBUTION.md in Apache APISIX repository)
### Installing Section (partial)
  - Prerequisites 
  - Different methods for installing Apache APISIX - RPM, DEB, Docker, Source Release
  - Building from the source
  - Stepwise procedures and solutions for possible traps and problems
## Apache APISIX configuration (missing)
  - What are the different types of configuration Apache APISIX needs?
  - Description of fields and their examples
  - List of supported/required Environment variables
  - First party/third party extensions for configuration files(VS Code etc.) 
  - Custom configuration and how to add them
  - Configuration examples
  - Sample Project (in later stages)
## Architectural Design (partial)
- Add a description to the diagram (current status: only diagrams no description)
- Remove configuration from this section
- Restructure this section to only contain architectural design
- Move definitions to "References" and if configuration then move to "Configuration"
  - Plugin 
  - Service
  - Consumer
  - Router
  - Script
  - Upstream
  - Route
## Plugins (need to be restructured)
- Introduction to Apache APISIX plugins
- How do they work?
- Getting started with plugins
- List of all plugins with links to their corresponding documentation
- Creating custom plugins
  - Prerequisites
  - How to create a custom plugin?
## Miscellaneous (missing)
- Possible sections
  - Support + Community
  - Extra Features

# Solution

## Here is the proposal for a new structure for the APISIX documentation. 
_(Sorry for the bad quality image 😓 )_
- Each checkbox represent a chapter/page
- Each bullet denotes *must-contain* or *preferred* headers in a page.
---
![image](https://user-images.githubusercontent.com/74113200/155853571-dba1adcb-7e0a-401e-b34a-d0c7b2141268.png)
---
# Checklist according to the proposal

- [ ] Welcome Page
    - Quick View
    - What is the latest in APISIX?
    - Featured
    - Changelogs
- [ ] Quickstart (APISIX in 5 mins)
    - Overview of APISIX
    - Setup and use APISIX right away
- [ ] Getting Started
    - [ ] Installation
        - From Package Managers
        - Docker Installation
        - Buiding from source
    - [ ] Contribution Guide `link to: https://apisix.apache.org/docs/general/contributor-guide`
        -  How to contribute
        - Type of contributions
        - Opening new issue
        - Opening a pull request
        - Code style
        - Claim your goodies (Yes we have contributor's t-shirt 👕 )
    - [ ] Community Guide
        - Slack
        - Mailing Lists
        - Other communities
        - Community Efforts
- [ ] Learning Library
    - [ ] Learn APISIX
        - YouTube Video Links
        - Tutorial Links (Katacoda + Written)
    - [ ] Usecases / How to guides
        - Can be used for katacoda tutorials
- [x] Concepts
    - [x] Route
    - [x] Plugin
    - [x] Script
    - [x] Service
    - [x] Consumer
    - [x] Upstream
    - [x] Global Rule
- [ ] Configuration
    - [ ] Introduction and types of configurations
    - [ ] List of fields and configuration examples
    - [ ] Supported environment variables
    - [ ] Custom Configurations
    - [ ] Scripts
    - [ ] Debug Mode
    - [ ] Plugin Configuration
- [ ] Architectural Design
    - [ ] APISIX Architecture
    - [ ] Plugin Architecture
        - Plugin loading process
        - Plugin Hierarchy
        - External Plugin Architecture
- [ ] API
    - [ ] Admin API
        - Split the pages 
    - [ ] Control API
        - Split the pages
- [ ] Plugins
    - [ ] Introduction to Apache APISIX plugin
    - [ ] List of all plugins with link to their description
        - One plugin per page
    - [ ] Plugin Developement
        - [ ] Setup Plugin Development Environement
            - Plugin Development Kit
        - [ ] Plugin Development Process
        - [ ] Writing Test Cases
        - [ ] Openresty/ngx_lua development guide (link if already exists)
    - [ ] External Plugins
    - [ ] Plugin configurations
    - [ ] Plugin Testing
- [ ] Integrations
    - [ ] What are integrations / extensions
    - [ ] Officially supported integrations/extensions
    - [ ] Community developed integrations / Third party plugins
- [ ] Miscellaneous
    - [ ] Support
        - [ ] Where to contact for help
        - [ ] Known issues and their solutions
        - [ ] How to report issue/bug
    - [ ] FAQ
    - [ ] Extra Features
        - [ ] Stand Alone Mode
        - [ ] Batch Processor
        - Add more here
    - [ ] Protocols
        - [ ] HTTP
        - [ ] GRPC
    - [ ] References
    - [ ] Heath Check
   
   ---
  ## Feel free to add missing parts, and DONOT mention this issue with closing keywords. This issue is strictly for discussion only. 
  ---
  _Currently this proposal is in the review phase, once approved by the community, you can pick up issues and start working on them_

**Comment by tokers** [comment]
Could the description of 'external plugins' more deliberated? I think there could have several documents to introduce the way to write plugins in different languages.

**Comment by nfrankel** [comment]
It depends what you mean by "little introduction". I really dislike coming to a project I don't know, and being presented with the update of the latest versions, but not understanding what it is about.

Good examples:

* https://heron.apache.org/
* https://kafka.apache.org/
* https://apisix.apache.org/

**Comment by juzhiyuan** [comment]
Hi, Avinal,

Thanks for your new structure 🤗

1. Getting Started → Contribution Guide/Community Guide: How about linking to [`https://apisix.apache.org/docs/general/contributor-guide`](https://apisix.apache.org/docs/general/contributor-guide)?
2. Usecases/How-to Guides: Are they the same contents?
3. API → AdminAPI: Will we split different entities into different sections?
4. Plugins → External Plugins: Just to make sure, do you mean Java/Go/xx Plugin Runner or plugins made by unofficial members? e.g [`https://github.com/swisscom/apisix-opa-plugin`](https://github.com/swisscom/apisix-opa-plugin)
5. Plugins: Many developers report that they cannot work well with Plugin Development in different machines/environments, could we have some recommended environments? or PDK.

Also cc @moonming @membphis to have a review, the new structure is important to us community and users.

**Comment by moonming** [comment]
@avinal great work 👍
Documentation is a part of Apache APISIX and is the most important part, looking forward to the new documentation structure

**Comment by shuaijinchao** [comment]
Great, in the new documentation structure, I see the logic of the documentation context and the process of helping users understand APISIX. Rather than feature points scattered across different documents like they are currently.

**Comment by tao12345666333** [comment]
:+1:  I think it's important to add "What is an API Gateway?"

**Comment by moonming** [comment]
We can add security, observability, Kubernetes deployment, multiple development language development plugins, WASM, etc. In the protocol part, MQTT and GraphQL are also worth mentioning.
Others LGTM

**Comment by leslie-tsang** [comment]
Would it be better to add a `Concepts` section to introduce `route` `service` `consumer` `upstream` ?

**Comment by juzhiyuan** [comment]
> It depends what you mean by "little introduction". I really dislike coming to a project I don't know, and being presented with the update of the latest versions, but not understanding what it is about.
> 
> Good examples:
> 
> * https://heron.apache.org/
> * https://kafka.apache.org/
> * https://apisix.apache.org/

Agree 🤗 @avinal we could pick some parts from current resources? like README, blogs to describe Apache APISIX?

**Comment by fbartels** [comment]
A very nice initiative. When getting started with Apisix I found myself constantly switching between documentation and blog posts around specific topics. Looks like your proposed "use cases" chapter would combine this quite nicely.

**Comment by tzssangglass** [comment]
> 5\. Plugins: Many developers report that they cannot work well with Plugin Development in different machines/environments, could we have some recommended environments? or PDK (Plugin Development Kit, @tzssangglass is familiar with this).

Perhaps a `Plugin Development` directory is needed to collect documentation related to plugin development.

under `Plugin Development` maybe:

1. Build development environment
2. Development Process
3. Development and debugging skills
4. Writing test cases

Now we have some documentation covering these.

This is a great proposal.

**Comment by tokers** [comment]
> > 5. Plugins: Many developers report that they cannot work well with Plugin Development in different machines/environments, could we have some recommended environments? or PDK (Plugin Development Kit, @tzssangglass is familiar with this).
> 
> Perhaps a `Plugin Development` directory is needed to collect documentation related to plugin development.
> 
> under `Plugin Development` maybe:
> 
> 1. Build development environment
> 2. Development Process
> 3. Development and debugging skills
> 4. Writing test cases
> 
> Now we have some documentation covering these.
> 
> This is a great proposal.

Besides these aspects, I hope that the below things can be added:

1. Guide about openresty/ngx_lua development (give some materials are also good);
2. Not only how to write case but also how to run them;
3. The Plugin Development Kit docs

**Comment by avinal** [comment]
Hi all, thank you all for your responses. Some of them are great suggestions and must be added. I am actively seeing all the responses and will add them in bulk 😃 . Thanks again 🙇🏼

**Comment by danielkocot** [comment]
I really like the idea of putting the documentation rewrite in to a community process. As @fbartels pointed out in the existing documentation you are mostly switching between the documentation and blog posts. It also might be good not to having a brief quick start which is normally relevant according to the use-cases interested users normally.

**Comment by nfrankel** [comment]
@danielkocot 

1. The first principle is that documentation is in sync with the code. On the other hand, blog posts naturally drift with time, and that's OK
2. The second principle is to use the [following documentation framework](https://documentation.divio.com/):

![](https://documentation.divio.com/_images/overview.png)

**Comment by Baoyuantop** [comment]
> > 5. Plugins: Many developers report that they cannot work well with Plugin Development in different machines/environments, could we have some recommended environments? or PDK (Plugin Development Kit, @tzssangglass is familiar with this).
> 
> Perhaps a `Plugin Development` directory is needed to collect documentation related to plugin development.
> 
> under `Plugin Development` maybe:
> 
> 1. Build development environment
> 2. Development Process
> 3. Development and debugging skills
> 4. Writing test cases
> 
> Now we have some documentation covering these.
> 
> This is a great proposal.

Strongly agree. This content is very necessary for new developers.

**Comment by guoqqqi** [comment]
> Learn APISIX
> * YouTube Video Links
> * Tutorial Links (Katacoda + Written)

Hi, I have a little advice here: 
- I think it would be a good idea to put the learning APISIX and user cases into a single place like a learning library.

**Comment by juzhiyuan** [comment]
> * Not only how to write case but also how to run them;

Many developers complain that it's hard to run **all** tests successful in local 😅

**Comment by moonming** [comment]
> > * Not only how to write case but also how to run them;
> 
> Many developers complain that it's hard to run **all** tests successful in local 😅

Running a complete collection of test cases for a modern open source project is an almost impossible task. The CI of Apache APISIX relies on Github actions, which contains a lot of docker and environment variable settings

**Comment by membphis** [comment]
> > * Not only how to write case but also how to run them;
> 
> Many developers complain that it's hard to run **all** tests successfully local 😅 AFAIK, it's related to many detailed things (may need @membphis @leslie-tsang's description), though it's hard, is it worth considering this to our tutorial?

Indeed, trying to run all test cases locally is not feasible. APISIX's test cases are already complex, so it is best for developers to run all test cases in Github Action CI.

**Comment by tao12345666333** [comment]
For the development environment, we can also give an example using GitPod [1] directly, or provide a development environment based on Docker images. The developer experience will be better.

[1] https://www.gitpod.io/

**Comment by leslie-tsang** [comment]
> Many developers complain that it's hard to run all tests successfully local 😅 AFAIK

Yep, the test case for APISIX is more complex than ever before. It will cause a hard time for newbie to walk though it.

> though it's hard, is it worth considering this to our tutorial?

IMO, It's worth setting up a simple test case ENV in the tutorial.

> Indeed, trying to run all test cases locally is not feasible

Yep, Especially after the introduction of the K8s discover module.

**Comment by juzhiyuan** [comment]
> > > * Not only how to write case but also how to run them;
> > 
> > 
> > Many developers complain that it's hard to run **all** tests successful in local 😅
> 
> Running a complete collection of test cases for a modern open source project is an almost impossible task. The CI of Apache APISIX relies on Github actions, which contains a lot of docker and environment variable settings

@moonming This is the current status, for developers, how do they know this clearly and easily? Our docs don't have a clear reminder (WARNING/INFO tip) for them, they have to know this fact when they faced problems and search through issues.

**Comment by anjia0532** [comment]
Maybe add `Best Practices` ? e.g. Different cluster size and best practices , Upgrade best practices , ide best practices(idea emmylua plugin, vscode+emmylua plugin , zbstudio ...),

**Comment by membphis** [comment]
> For the development environment, we can also give an example using GitPod [1] directly, or provide a development environment based on Docker images. The developer experience will be better.
> 
> [1] https://www.gitpod.io/

@nic-6443 can we run the Github Action CI in gitpod?

**Comment by nic-6443** [comment]
> > For the development environment, we can also give an example using GitPod [1] directly, or provide a development environment based on Docker images. The developer experience will be better.
> > [1] https://www.gitpod.io/
> 
> @nic-6443 can we run the Github Action CI in gitpod?

Gitpod is an ubuntu environment, so we can build a development environment in Gitpod to run APISIX test cases. Gitpod is not a suitable runtime environment for Github Action, which is a CI-oriented product, but Gitpod is a developer-oriented product.

**Comment by avinal** [comment]
Hello all, Thank you all for the review. here is the update for the proposal based on your suggestions. 

> Could the description of 'external plugins' more deliberated? I think there could have several documents to introduce the way to write plugins in different languages.

 Yeah sure, there is just one problem, since external plugins may or may not be supported by APISIX, we can only explain the preferred way to write them and not explain every possible way. It would be good to have different pages for different writing methods.

 
> It depends what you mean by "little introduction". I really dislike coming to a project I don't know, and being presented with the update of the latest versions, but not understanding what it is about.

Little introduction means just necessary stuff about the project, neither too deep nor too shallow. I will add some explanation. 
Thank you for the examples, will add something similar to this doc as well.



> 1. Getting Started → Contribution Guide/Community Guide: How about linking to [`https://apisix.apache.org/docs/general/contributor-guide`](https://apisix.apache.org/docs/general/contributor-guide)?

Both ways are possible, either we can move the whole doc to the contributor guide section or just link it. Since the doc is already there, it would be nice to just link.

> 2. Usecases/How-to Guides: Are they the same contents?

There is a slight difference between these two topics. The use cases are what we intend the application to be used for. And how-to Guides are questions answered from a particular use-case perspective. In general, they mean the same. I put it there just to decide which topic to use.

> 3. API → AdminAPI: Will we split different entities into different sections?

If there are separable entities then it makes sense to split them into different sections. Right now the page is just too large, ideally, one page shouldn't contain more than 2000 words(excluding code), or it becomes too hard to follow and search. So splitting will be nice.

> 4. Plugins → External Plugins: Just to make sure, do you mean Java/Go/xx Plugin Runner or plugins made by unofficial members? e.g [`https://github.com/swisscom/apisix-opa-plugin`](https://github.com/swisscom/apisix-opa-plugin)

Yes, we can list popular external plugins and the preferred method to create them. This will expand the applications of APISIX.

> 5. Plugins: Many developers report that they cannot work well with Plugin Development in different machines/environments, could we have some recommended environments? or PDK (Plugin Development Kit, @tzssangglass is familiar with this).

Yes, that is why we are going to a plugin development section. And we can then suggest the best practices for plugin development.


> We can add security, observability, Kubernetes deployment, multiple development language development plugins, WASM, etc. In the protocol part, MQTT and GraphQL are also worth mentioning. 

Sure, I am not very familiar with the protocols, will add them once I have discussed them with the engineers. 


> Would it be better to add a `Concepts` section to introduce `route` `service` `consumer` `upstream` ?

Yeah, I have already added that `Definition of important keywords` inside `architectural design`, but `Concepts` seems a better choice of word, please suggest if I should move them out in a separate chapter instead.


> Perhaps a `Plugin Development` directory is needed to collect documentation related to plugin development.
> 
> under `Plugin Development` maybe:
> 
>     1. Build development environment
> 
>     2. Development Process
> 
>     3. Development and debugging skills
> 
>     4. Writing test cases

I have added this as `Creating custom plugins`, but `Plugin Development` seems more suitable, will rename it and add the suggested sections. 


> Besides these aspects, I hope that the below things can be added:
> 
>     1. Guide about openresty/ngx_lua development (give some materials are also good);
> 
>     2. Not only how to write case but also how to run them;
> 
>     3. The Plugin Development Kit docs

Seems like a great addition to me. Will add them as well.

> Hi, I have a little advice here:
> 
>     * I think it would be a good idea to put the learning APISIX and user cases into a single place like a learning library.

That would be more clear and expressive, Will restructure.

> Maybe add `Best Practices` ? e.g. Different cluster size and best practices , Upgrade best practices , ide best practices(idea emmylua plugin, vscode+emmylua plugin , zbstudio ...).
> 
> e.g. ansible best practices https://docs.ansible.com/ansible/2.8/user_guide/playbooks_best_practices.html

There was some consideration at first but haven't thought about it much. I will see them and add them. Thanks for the example.

**Comment by juzhiyuan** [comment]
LGTM

**Comment by tzssangglass** [comment]
LGTM

**Comment by anjia0532** [comment]
Maybe create `aswsome apisix` repo and add to docs(outer link)? e.g. <https://github.com/bungle/awesome-resty>,<https://github.com/posquit0/awesome-kong> @juzhiyuan @avinal

**Comment by avinal** [comment]
> Maybe create `aswsome apisix` repo and add to docs(outer link)? e.g. https://github.com/bungle/awesome-resty,https://github.com/posquit0/awesome-kong @juzhiyuan @avinal

I like the idea, we can then add it to the Awesome lists. Cool 👍🏼

**Comment by hf400159** [comment]
I think Apache APISIX should provide a learning roadmap in the community. There are two RoadMaps for different users, one is for those who have not been exposed to API Gateway, and the other is for those who have already used API Gateway. This allows users to quickly learn Apache APISIX.

**Comment by leslie-tsang** [comment]
LGTM

**Comment by navendu-pottekkat** [comment]
This issue has become stale and might need to be revisited on a smaller scale. People looking to work on parts of this issue, please consider it stale and closed. We will open a new issue with an updated proposal.
